js/hosts/pins.browser.js

```diff
@@ -2,6 +2,9 @@
0?v(l(s,5),(e=h)&(o=a)^e&(i=f)^o&i,c,2400959708,u[p]):v(l(s,5),h^a^f,c,3395469782,u[p]),c=f,f=a,a=l(h,30),h=s,s=n;return r[0]=w(s,r[0]),r[1]=w(h,r[1]),r[2]=w(a,r[2]),r[3]=w(f,r[3]),r[4]=w(c,r[4]),r}function p(t,r,n,e){for(var o,i=15+(r+65>>>9<<4),u=r+n;t.length<=i;)t.push(0);for(t[r>>>5]|=128<<24-r%32,t[i]=4294967295&u,t[i-1]=u/4294967296|0,o=0;o<t.length;o+=16)e=c(t.slice(o,o+16),e);return e}return A.prototype.update=function(t){for(var r=0,n=this.T>>>5,e=this.s(t,this.Y,this.I),o=e.binLen,i=e.value,u=o>>>5,s=0;s<u;s+=n)r+this.T<=o&&(this.U=this.A(i.slice(s,s+n),this.U),r+=this.T);this.H+=r,this.Y=i.slice(r>>>5),this.I=o%this.T,this.C=!0},A.prototype.getHash=function(t,r){var n,e,o=this.m,i=a(r);if(this.F){if(-1===i.outputLen)throw new Error("Output length must be specified in options");o=i.outputLen}var u=h(t,o,this.u,i);if(this.L&&this.i)return u(this.i(i));for(e=this.R(this.Y.slice(),this.I,this.H,this.p(this.U),o),n=1;n<this.numRounds;n+=1)this.F&&o%32!=0&&(e[e.length-1]&=16777215>>>24-o%32),e=this.R(e,o,0,this.l(this.g),o);return u(e)},A.prototype.setHMACKey=function(t,r,n){if(!this.t)throw new Error("Variant does not support HMAC");if(this.C)throw new Error("Cannot set MAC key after calling update");var e=u(r,(n||{}).encoding||"UTF8",this.u);this.B(e(t))},A.prototype.B=function(t){var r,n=this.T>>>3,e=n/4-1;if(1!==this.numRounds)throw new Error("Cannot set numRounds with MAC");if(this.L)throw new Error("MAC key already set");for(n<t.binLen/8&&(t.value=this.R(t.value,t.binLen,0,this.l(this.g),this.m));t.value.length<=e;)t.value.push(0);for(r=0;r<=e;r+=1)this.N[r]=909522486^t.value[r],this.S[r]=1549556828^t.value[r];this.U=this.A(this.N,this.U),this.H=this.T,this.L=!0},A.prototype.getHMAC=function(t,r){var n=a(r);return h(t,this.m,this.u,n)(this.o())},A.prototype.o=function(){if(!this.L)throw new Error("Cannot call getHMAC without first setting MAC key");var t=this.R(this.Y.slice(),this.I,this.H,this.p(this.U),this.m),r=this.A(this.S,this.l(this.g));return this.R(t,this.m,this.T,r,this.m)},n(t=e,r=i=A),t.prototype=null===r?Object.create(r):(o.prototype=r.prototype,new o),e;function e(t,r,n){var e=this;if("SHA-1"!==t)throw new Error("Chosen SHA variant is not supported");var o=n||{};return(e=i.call(this,t,r,n)||this).t=!0,e.i=e.o,e.u=-1,e.s=u(e.h,e.v,e.u),e.A=c,e.p=function(t){return t.slice()},e.l=f,e.R=p,e.U=[1732584193,4023233417,2562383102,271733878,3285377520],e.T=512,e.m=160,e.F=!1,o.hmacKey&&e.B(function(t,r){var n="hmacKey must include a value and format";if(!t)throw new Error(n);if(void 0===t.value||!t.format)throw new Error(n);return u(t.format,t.encoding||"UTF8",r)(t.value)}(o.hmacKey,e.u)),e}function o(){this.constructor=t}function A(t,r,n){var e=n||{};if(this.h=r,this.v=e.encoding||"UTF8",this.numRounds=e.numRounds||1,isNaN(this.numRounds)||this.numRounds!==parseInt(this.numRounds,10)||this.numRounds<1)throw new Error("numRounds must a integer >= 1");this.g=t,this.Y=[],this.I=0,this.C=!1,this.H=0,this.L=!1,this.N=[],this.S=[]}});
```

Comment:
I was wondering: why not to add sha1 script to utils folder and use it without a special callback? In this case we can get rid of crypto.

Reply:
Ah, the problem is that computing a `sha1` is probably quite slow if we do it from JavaScript while running in Python/R. JavaScript is quite fast so the sha1 is fine from the browser or nodejs, but since R/Python will run a not-so-fast JS interpreter, I'm almost sure this will become painfully slow. So we need to do all the expensive tasks in R/Python/JS callbacks whenever possible.